### PR TITLE
feat(data): add partial vector data structure

### PR DESCRIPTION
**Pull request by vapourismo**
Part of RV-861

# What

Introduces a partial vector data structure that lets you define arbitrary parts of a large space (`0..=usize::MAX`).

# Why

I found this data structure to be very useful in implementing the byte array state component (#732). It was also handy in making `DataSpace` more space efficient during proof generation (#736).

# How

I added lots of commentary within the data structure module. Please let me know what you think is missing.

# Manually Testing

```
make all
```

# Tasks for the Author

- [x] Link all Linear issues related to this MR using magic words (e.g. part of, relates to, closes).
- [x] Eliminate dead code and other spurious artefacts introduced in your changes.
- [x] Document new public functions, methods and types.
- [x] Make sure the documentation for updated functions, methods, and types is correct.
- [x] Add tests for bugs that have been fixed.
- [x] [Explain changes](#regressions) to regression test captures when applicable.
- [x] Write commit messages in agreement with our guidelines.
- [x] Self-review your changes to ensure they are high-quality.
- [x] Complete all of the above before assigning this MR to reviewers.

### CODE REVIEW

**Comment by vapourismo** [comment]
* **#731** <a href="https://app.graphite.com/github/pr/tezos/riscv-pvm/731?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#732](https://github.com/tezos/riscv-pvm/pull/732) <a href="https://app.graphite.com/github/pr/tezos/riscv-pvm/732?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#736](https://github.com/tezos/riscv-pvm/pull/736) <a href="https://app.graphite.com/github/pr/tezos/riscv-pvm/736?utm_source=stack-comment-icon" target="_blank">Graphite</a>) 👈 <a href="https://app.graphite.com/github/pr/tezos/riscv-pvm/731?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: data/src/partial_vec.rs [hunk 1]
```diff
@@ -0,0 +1,441 @@
...
+                    // needs to be kept.
+                    if undefined_suffix_len > 0 {
+                        let suffix = Self::undefined(undefined_suffix_len);
+                        new_data_node = new_data_node.concat(suffix);
+                    }
+
+                    *target = new_data_node;
+                }
+
+                PartialVec::Defined { data } => {
```
Comment (NSant215) [inline data/src/partial_vec.rs]:
```suggestion
                Self::Defined { data } => {
```

File: data/src/partial_vec.rs [hunk 2]
```diff
@@ -0,0 +1,441 @@
+// SPDX-FileCopyrightText: 2025 TriliTech <contact@trili.tech>
```
[the remark this answers is not in the document]
Comment (vapourismo) [reply data/src/partial_vec.rs]: Fixed in this [diff](https://github.com/tezos/riscv-pvm/compare/98b80a564a87cf3a9ad58a911d254225ee4684ea..2e5e7ed2a550fb9a789c4c9c0b59b438daa58ee5).

File: data/src/partial_vec.rs [hunk 3]
```diff
@@ -0,0 +1,441 @@
...
+                        *target = Self::defined(data).concat(new_node);
+
+                        continue;
+                    }
+
+                    let new_data_end = offset + new_data.len();
+
+                    // When we start inserting within the defined range, but the new data extends
+                    // beyond the existing range, we can truncate and append in place.
+                    if new_data_end >= data.len() {
```
[the remark this answers is not in the document]
Comment (vapourismo) [reply data/src/partial_vec.rs]: I added a bit more commentary to reference that the `Concatenated` case gives us these guarantees.

**Review by vapourismo — COMMENTED** [review]
(GitHub is kidnapping my comments, so I had to send them via the "review" function).

File: data/src/partial_vec/tests.rs [hunk 4]
```diff
@@ -0,0 +1,507 @@
+// SPDX-FileCopyrightText: 2026 TriliTech <contact@trili.tech>
+//
+// SPDX-License-Identifier: MIT
+
+//! Tests for [`PartialVec]
```
Comment (thomasathorne) [inline data/src/partial_vec/tests.rs]: Super nit, I don't think this bit of doc will be generated anyway?

```suggestion
//! Tests for [`PartialVec`]
```

File: data/src/partial_vec/tests.rs [hunk 5]
```diff
@@ -0,0 +1,507 @@
...
+    // After truncating to 0, the vector should have no defined data
+    assert!(!vec.is_partially_defined(0..5));
+}
+
+/// Check that truncating an undefined vector works correctly.
+#[test]
+fn truncate_undefined_vector() {
+    // BUG: When truncating an Undefined node, the code does `length -= *undefined_length`
+    // which causes underflow when *undefined_length > length.
+    // It should instead set *undefined_length = length.
```
Comment (kurtisc) [inline data/src/partial_vec/tests.rs]: Are these tests that these bugs have been fixed? The comments make it seem like this is what they do, rather than what they did.

File: data/src/partial_vec/tests.rs [hunk 6]
```diff
@@ -0,0 +1,507 @@
...
+        .copied()
+        .collect::<Vec<_>>();
+    assert_eq!(values, vec![1, 2, 3]);
+}
+
+/// Check that truncating to zero on a defined vector works correctly.
+#[test]
+fn truncate_to_zero_defined() {
+    // BUG: Truncating to 0 should clear the vector, but the loop condition
+    // `while length > 0` prevents this from happening.
```
Comment (kurtisc) [inline data/src/partial_vec/tests.rs]: See comment on `truncate_undefined_vector`

File: data/src/partial_vec.rs [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+    }
+
+    /// Create a vector of undefined data with the provided length.
+    pub fn undefined(length: usize) -> Self {
+        Self::Undefined { length }
+    }
+
+    /// Concatenation of two partial vectors into a new one.
+    fn concat(self, other: Self) -> Self {
+        Self::Concatenated {
```
Comment (victor-dumitrescu) [inline data/src/partial_vec.rs]: When concatenating 2 partial vectors of the same type, why not create a new partial vector of that type instead of a concatenation? I assume it's for proof safety, to avoid having large copies in one step?

File: data/src/partial_vec.rs [hunk 8]
```diff
@@ -0,0 +1,448 @@
...
+            while !range.is_empty() {
+                let Some(node) = work.pop_front() else {
+                    // If we ran out of things to do, but the range is not empty yet, then the partial
+                    // vector is undefined for the rest of the queried range.
+                    let entry = RangeEntry::Undefined {
+                        length: range.len(),
+                    };
+
+                    // We must drain the range to prevent further iterations.
+                    range = 0..0;
```
Comment (victor-dumitrescu) [inline data/src/partial_vec.rs]: Nit, could use `Range::default()`

**Review by NSant215 — COMMENTED** [review]
Generally LGTM, just wanna sanity check against test coverage as I can't see these cases covered but could easily be missing them:
- defining beyond current width of a vec but overlapping 
- `defined_range` querying a range with no defined ranges.

Also is there a reason you went for the name `PartialVec` rather than `SparseVec`?

File: data/src/partial_vec.rs [hunk 9]
```diff
@@ -0,0 +1,448 @@
...
+        loop {
+            // SAFETY: Dereferencing `target` is safe because we have not aliased the pointer.
+            match unsafe { &mut *target } {
+                Self::Undefined { length } => {
+                    *length = keep_length.min(*length);
+                    break;
+                }
+
+                Self::Defined { data } => {
+                    data.truncate(keep_length);
```
Comment (danielpapp-trilitech) [inline data/src/partial_vec.rs]: Is being "undefined" implicit for the piece that we cut off here?

File: data/src/partial_vec.rs [hunk 10]
```diff
@@ -0,0 +1,448 @@
...
+    pub fn truncate(&mut self, mut keep_length: usize) {
+        // The borrow checker is unfortunately a little in the way in this method. Please see the
+        // `PartialVec::Concatenated` case below for more information.
+        let mut target: *mut Self = self;
+
+        loop {
+            // SAFETY: Dereferencing `target` is safe because we have not aliased the pointer.
+            match unsafe { &mut *target } {
+                Self::Undefined { length } => {
+                    *length = keep_length.min(*length);
```
Comment (danielpapp-trilitech) [inline data/src/partial_vec.rs]: Why are we cutting here? I mean this range is already undefined, so matching on this should be a no-op, no?

File: data/src/partial_vec.rs [hunk 11]
```diff
@@ -0,0 +1,448 @@
...
+/// Partial vector that maybe not be defined for certain ranges
+///
+/// This data type can be seen as a `BTreeMap<usize, T>` which provides a more memory-efficient
+/// representation for when there are many adjacent keys and the primary query mechanism is using
+/// ranges.
+#[derive(Debug, Clone)]
+pub enum PartialVec<T> {
+    /// Undefined vector of a certain width
+    Undefined {
+        /// How many bytes are undefined
```
Comment (thomasathorne) [inline data/src/partial_vec.rs]: This was unintuitive (to me at least) so I'd add something emphasising it is bytes and _not_ elements (like you've done below for `node_width`).

**Review by thomasathorne — COMMENTED** [review]
Looks nice so far, a few comments. Will keep reviewing.

File: data/src/partial_vec.rs (hunk 9 above)
Comment (vapourismo) [reply data/src/partial_vec.rs]: Yes.

File: data/src/partial_vec.rs [hunk 12]
```diff
@@ -0,0 +1,448 @@
...
+        self.range(range).filter_map(move |chunk| match chunk {
+            RangeEntry::Undefined { length } => {
+                start += length;
+                None
+            }
+
+            RangeEntry::Defined { data } => {
+                let new_start = start + data.len();
+                let item = (start, data);
+                start = new_start;
```
[the remark this answers is not in the document]
Comment (vapourismo) [reply data/src/partial_vec.rs]: Argh no! That's an artefact from a previous version. I applied your suggestion.

File: data/src/partial_vec.rs (hunk 7 above)
Comment (vapourismo) [reply data/src/partial_vec.rs]: It's possible to do that. I hadn't added it until now because that case never happens.

I added it now, given it might be useful in the future. I'm hoping the inlining will take care of preventing the additional cycles spend on checking the variants.

File: data/src/partial_vec.rs (hunk 10 above)
Comment (vapourismo) [reply data/src/partial_vec.rs]: I tried it without updating the length. There are enough indirect guardrails in place that this will currently not break things.

My initial assessment was this:

image

If we don't update the length, this will happen:

image

The data structure itself is in an incoherent state. However! The current guardrails seem to make the data structure converge towards a coherent state - at least for the operations I've tried.

I don't fully trust this yet, though. So I'd prefer to update the length of undefined nodes as a matter of precaution.

**Review by NSant215 — APPROVED** [review]

**Review by thomasathorne — CHANGES_REQUESTED** [review]
This generally looks great, the PR was fun to review!

My one concern is that there is nothing keeping the tree balanced. So for example
```
#[test]
fn unbalanced_tree() {
    let mut vec: PartialVec<u8> = PartialVec::undefined(0);

    for i in 0..40000 {
        vec.define(i * 2 + 1, vec![0]);
    }
}
```
is enough to fail with a stack overflow.

That is obviously a long way from the intended use case, but 40,000 entries isn't _that_ many. I'd argue we should either consider fixing this or be very careful not to use this in any situations where an adversary can choose the shape of the data.

File: data/src/partial_vec/tests.rs [hunk 13]
```diff
@@ -0,0 +1,510 @@
...
+    // Now create a structure with defined regions on both sides
+    vec.define(0, vec![1, 2, 3]); // Left side: indices 0-2
+    vec.define(7, vec![7, 8, 9]); // Right side: indices 7-9
+
+    // Now define data that spans across the middle, overlapping both regions
+    // This should trigger the cross-boundary insertion code path
+    vec.define(2, vec![20, 30, 40, 50, 60]);
+
+    // Verify the result: indices 0-1 should be [1, 2], indices 2-6 should be [20, 30, 40, 50, 60]
+    assert!(vec.is_any_defined(0..7));
```
Comment (thomasathorne) [inline data/src/partial_vec/tests.rs]: This line isn't actually checking what the comment above claims.

**Review by emturner — COMMENTED** [review]
overall makes sense, main comment is that I'd argue a significant number of edge cases seem to exist w.r.t. integer overflow that should be guarded against

**Review by emturner — APPROVED** [review]

**Review by thomasathorne — APPROVED** [review]
Happy to approve this now and look forward to the follow-up.

File: data/src/partial_vec.rs [hunk 14]
```diff
@@ -0,0 +1,441 @@
...
+                        continue;
+                    }
+
+                    // At this point we know that the inserted range overlaps with both the left and
+                    // right nodes. So we divide the insertion vector accordingly.
+                    let left_overlap_tail = left.width() - offset;
+                    let new_right_data = new_data.split_off(left_overlap_tail);
+
+                    work.push((left, offset, new_data));
+                    work.push((right, 0, new_right_data));
```
[the remark this answers is not in the document]
Comment (vifar) [reply data/src/partial_vec.rs]: Fixed in commit fb772c530. price2 is now at the top level of the orderInfo type in both helpers and accessed via orderInfo.price2.
